### PR TITLE
Parse dates from Pubmed records into pub_hash

### LINKED ISSUE #1441: Map dates for Pubmed harvested records in pub hash
For WoS and ORCID, we currently map the date into the pub_hash, but we don't do this Pubmed soured records.  

See https://github.com/sul-dlss/sul_pub/blob/main/app/models/pubmed_source_record.rb#L109 for where the Pubmed records are mapped to pub hash.

See https://github.com/sul-dlss/sul_pub/blob/main/lib/web_of_science/map_citation.rb#L28 for where the date is mapped in WoS records

See https://github.com/sul-dlss/sul_pub/blob/main/lib/orcid/work_mapper.rb#L31 for where the date is mapped in ORCID records

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made?

Fixes #1441 (upcoming request from Profiles):  if available, we should map dates from PubmedSource Records into the same YYYY-MM-DD format as we do for WoS and ORCID records

Notes:
- refactored how the year (currently done in our code) is extracted so we can use the same type of code for the month/year
- could add some more Pubmed fixtures to test some more permutations of date locations/formats to be parsed
- once done, we will need to go back and rebuild the pubhash for Pubmed sourced records if we want to the date added

There is a rake task for rebuilding Pubmed pub hashes from a previous remediation, but it starts with cap_profile_ids: see https://github.com/sul-dlss/sul_pub/blob/main/lib/tasks/pubmed.rake#L7  In this case, we may want to do *all* Pubmed sourced publications (or all retrieved after a certain date), so we'd have a much bigger set to start from.  Basically all you have to do is call `publication#rebuild_pub_hash` on any publication that has a Pubmed provenance to remap.  See https://github.com/sul-dlss/sul_pub/issues/1441#issuecomment-1054842850 for how to do this (tested in cap-dev)

Note that rebuilding pubhashes will touch the "last_updated" field in the publication record, which will then cause the publication to be returned to Profiles as updated in their nightly update API call.  So after we do the remediation (if we do it), they will end up with a huge batch of updated publications (when they are used to a much smaller number).  Edit: there will be ~17910 (see https://github.com/sul-dlss/sul_pub/issues/1441#issuecomment-1054842850)

See https://github.com/sul-dlss/sul_pub/wiki/Useful-console-commands-(harvest,-WoS,-Pubmed,-others)#provenance-publication-stats for some ideas on how to query for the Pubmed provenance publications.

## How was this change tested?

- Added some new tests
- Deployed to cap-dev, which has ~16k publications that are Pubmed provenance.  Added the date to all of these and then verified that the date's year matched the year using https://github.com/sul-dlss/sul_pub/issues/1441#issuecomment-1054842850

### CODE REVIEW

**Comment by peetucket** [comment]
See https://github.com/sul-dlss/sul_pub/issues/1441#issuecomment-1054842850 for an example of how to remediate the previous harvest pubmed source dates

File: app/models/pubmed_source_record.rb [hunk 1]
```diff
@@ -99,6 +99,61 @@ def extract_mesh_headings_from_pubmed_record(pubmed_record)
...
+    pubmed_date_xpaths('Year').each do |path|
+      match = publication.xpath(path).text.match(/[12][0-9]{3}/)
+      next unless match
+
+      year = match.to_s
+      break
+    end
+
+    year
+  end
```
Comment (mjgiarlo) [inline app/models/pubmed_source_record.rb:117]: @peetucket I wonder if it's worth simplifying this method (and its buddies below) to something more like:

```suggestion
potential_years.find { |year| year.match?(/[12][0-9]{3}/) }.to_s
  # see https://dtd.nlm.nih.gov/ncbi/pubmed/doc/out/180101/el-Year.html
  def extract_year_from_pubmed_record(publication)
    pubmed_date_xpaths('Year').find { |path| publication.xpath(path).text.match?(/[12][0-9]{3}/) }.to_s.presence
  end
```

Comment (peetucket) [reply app/models/pubmed_source_record.rb]: Thanks for the suggestion.  I tweaked slightly to make it work (this wasn't quite working as expected), and applied a similar change to the other day and month methods.

File: app/models/pubmed_source_record.rb [hunk 2]
```diff
@@ -139,22 +171,16 @@ def source_as_hash(publication = Nokogiri::XML(source_data).xpath('//PubmedArtic
...
+    year = extract_year_from_pubmed_record(publication)
+    record_as_hash[:year] = year if year
+
+    month = extract_month_from_pubmed_record(publication)
+    day = extract_day_from_pubmed_record(publication)
+    if year && month && day
+      record_as_hash[:date] =  "#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')}"
+    elsif year && month
+      record_as_hash[:date] =  "#{year}-#{month.rjust(2, '0')}"
     end
```
Comment (peetucket) [inline app/models/pubmed_source_record.rb:184]: note, we want to 0 pad months and days to be consistent with other date formats, so 2022-1-4 becomes 2022-01-04

**Review by mjgiarlo — APPROVED** [review]

**Comment by peetucket** [comment]
@justinlittman : just pinging you in case you wanted to have a look and this before I merge ... since it was pretty straightforward, I think I'm going to move on to refactoring this a bit more to pull out the Pubmed pubhash generation into it's own class like we do for WoS and ORCID (instead of having it in the model "source record" class)

**Review by justinlittman — APPROVED** [review]

**Pull request merged**